### PR TITLE
Feat(browser emulator) Improve error stacks and a lot more

### PR DESCRIPTION
**Pull request by soundofspace**

### CODE REVIEW

File: plugins/default-browser-emulator/injected-scripts/SharedWorker.prototype.ts [hunk 1]
```diff
@@ -1,20 +1,73 @@
...
+  if (waiter in self) {
+    // trigger promise
+    self[waiter]();
+  } else {
+    // injected script fast enough -> no wait needed
+    self[waiter] = true;
+  }
 }
+
+if (typeof SharedWorker === 'undefined') {
```
Comment (blakebyrnes) [inline plugins/default-browser-emulator/injected-scripts/SharedWorker.prototype.ts:21]: should this come before the waiter? Not sure if we need this clause or not, but if it's possible, it seems that we would be stranding the wait- variables

File: plugins/default-browser-emulator/injected-scripts/SharedWorker.prototype.ts [hunk 2]
```diff
@@ -1,20 +1,73 @@
...
+// When launching a SharedWorker from blob it is possible that it launches so fast
+// that setting debugger breakpoint is too slow and not triggered. We handle this in two ways:
+// 1. SharedWorker blob is too fast -> create promise in blob and wait for injected logic to trigger promise resolver
+// 2. Debugger is fast enough -> set bool = true in injected logic -> sharedworker sees this and doesn't wait
+
+
+// Random variable so external scripts can't predict this
+const waiter = `wait-${sourceUrl}`;
+
+if (targetType === 'shared_worker') {
```
Comment (blakebyrnes) [inline plugins/default-browser-emulator/injected-scripts/SharedWorker.prototype.ts:11]: I think you might need one of these "per shared worker", won't you? If there are two waiting at once, wouldn't the variables not work?

File: plugins/default-browser-emulator/injected-scripts/_proxyUtils.ts [hunk 3]
```diff
@@ -31,66 +32,77 @@ const ObjectCached = {
   getOwnPropertyDescriptor: Object.getOwnPropertyDescriptor.bind(Object),
 };
 
+// Store External proxies as undefined so we can treat it as just missing
+// TODO we need to cleanup this thing or it will get out of control.
```
Comment (blakebyrnes) [inline plugins/default-browser-emulator/injected-scripts/_proxyUtils.ts:36]: Could we put WeakRefs into value?

File: plugins/default-browser-emulator/injected-scripts/_proxyUtils.ts [hunk 4]
```diff
@@ -192,97 +128,112 @@ function proxyConstructor<T, K extends keyof T>(
...
+    let protoTarget = newPrototype;
+    let newPrototypeProto;
+    try {
+      newPrototypeProto = Object.getPrototypeOf(newPrototype);
+    } catch {}
+    if (newPrototype === proxy || newPrototypeProto === proxy) {
+      protoTarget = target;
+    }
+
+    const temp = { stack: 'TODO' };
```
Comment (blakebyrnes) [inline plugins/default-browser-emulator/injected-scripts/_proxyUtils.ts:169]: Can we throw a different word in here so it doesn't autocomplete to project "todos"?

File: plugins/default-browser-emulator/injected-scripts/navigator.ts [hunk 5]
```diff
@@ -14,41 +14,29 @@ if ('NetworkInformation' in self) {
...
-      try {
-        // check if these work
-        await ReflectCached.apply(target, thisArg, argArray);
-      } catch (e) {
-        error = e;
-      }
-    }
-    if (error) return Promise.reject(cleanErrorStack(error));
+    checkThisArg(thisArg, "Failed to execute 'getHighEntropyValues' on 'NavigatorUAData': ");
+    // check if these work
```
Comment (blakebyrnes) [inline plugins/default-browser-emulator/injected-scripts/navigator.ts:37]: Is this a todo?

**Review by blakebyrnes — COMMENTED** [review]
Great stuff! Just a few comments/questions in here.

File: plugins/default-browser-emulator/injected-scripts/navigator.ts (hunk 5 above)
Comment (soundofspace) [reply plugins/default-browser-emulator/injected-scripts/navigator.ts:37]: This was already there I think so

**Pull request merged**